src/cart/crt_hlc.c

```diff
@@ -61,6 +61,31 @@ uint64_t crt_hlc_get(void)
 	return ret;
 }
 
+typedef void
+(*crt_hlc_error_cb) (void *arg);
+
+int
+crt_register_hlc_error_cb(crt_hlc_error_cb event_handler, void *arg)
+{
+	int rc = 0;
+
+	D_MUTEX_LOCK(&crt_plugin_gdata.cpg_mutex);
+	crt_plugin_gdata.hlc_error_cb = event_handler;
+	crt_plugin_gdata.hlc_error_cb_arg = arg;
+	D_MUTEX_UNLOCK(&crt_plugin_gdata.cpg_mutex);
+
+	return rc;
+}
+
+void
+crt_trigger_hlc_error_cb(void)
+{
+	if (crt_plugin_gdata.hlc_error_cb) {
+		crt_plugin_gdata.hlc_error_cb(
+					crt_plugin_gdata.hlc_error_cb_arg);
+	}
+}
```

Comment:
Sorry, missed this in my first review: Not taking cpg_mutex is unsafe in theory. A thread may be executing line 73 and line 74 above, while another thread executes line 83 and 84. In theory, is it possible that the latter sees the cb but not the arg?

Either we take the lock here in crt_trigger_hlc_error_cb, or we must instruct users to register the cb while there can be no other thread triggering crt_trigger_hlc_error_cb.

[Optional] This new mechanism can be outside of crt_hlc.c, for crt_hlc.c doesn't need to know anything about crt data structures. And the error reporting is not part of the HLC algorithm. Just a personal opinion.

Reply:
fixed